File: go/medium/1838_frequency_of_the_most_frequent_element.go
package medium

import (
	"sort"
)

func maxFrequency(nums []int, k int) int {
	n := len(nums)

	if n < 2 {
		return n
	}
	sort.Ints(nums)
	l, max_freq := 0, 1

	for r := 1; r < n; r++ {
		ops := (nums[r] - nums[r-1]) * (r - l)

		for l < r && k < ops {
			ops -= nums[r] - nums[l]
			l += 1
		}
		k -= ops
		max_freq = max(max_freq, r-l+1)
	}

	return max_freq
}

File: python/medium/1501-2000/1838_frequency_of_the_most_frequent_element.py
from typing import List


class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        n = len(nums)

        if n < 2:
            return n

        nums.sort()
        max_freq, l = 1, 0

        for r, x in enumerate(nums[1:], start=1):
            extra_ops = (x - nums[r - 1]) * (r - l)

            while k < extra_ops:
                extra_ops -= x - nums[l]
                l += 1

            k -= extra_ops
            max_freq = max(max_freq, r - l + 1)

        return max_freq
